system/lib/wasmfs/file_table.h

```diff
@@ -0,0 +1,101 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+// This file defines the open file table of the new file system.
+// Current Status: Work in Progress.
+// See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#pragma once
+
+#include "file.h"
+#include <assert.h>
+#include <mutex>
+#include <utility>
+#include <vector>
+#include <wasi/api.h>
+
+class OpenFileDescriptor {
+  std::shared_ptr<File> file;
+  __wasi_filedelta_t offset;
+  std::mutex mutex;
+
+public:
+  OpenFileDescriptor(uint32_t offset, std::shared_ptr<File> file);
+
+  std::shared_ptr<File>& getFile();
+};
+
+class FileTable {
+  static std::vector<std::shared_ptr<OpenFileDescriptor>> entries;
+  std::mutex mutex;
+
+  FileTable() {
+    entries.push_back(std::make_shared<OpenFileDescriptor>(0, std::make_shared<StdinFile>()));
+    entries.push_back(std::make_shared<OpenFileDescriptor>(0, std::make_shared<StdoutFile>()));
+    entries.push_back(std::make_shared<OpenFileDescriptor>(0, std::make_shared<StderrFile>()));
+  }
+
+public:
+  // Handle represents an RAII wrapper object. Access to the FileTable must go through a Handle.
+  // A Handle holds FileTable's lock for the duration of its lifetime.
+  class Handle {
+    FileTable& fileTable;
+    std::unique_lock<std::mutex> lock;
+
+  public:
+    Handle(FileTable& fileTable) : fileTable(fileTable), lock(fileTable.mutex) {}
+
+    // Returns size of the associated FileTable entries.
+    size_t size() { return fileTable.entries.size(); }
+
+    // Adds given OpenFileDescriptor to FileTable entries. Returns fd (insertion index in entries).
+    __wasi_fd_t addOpenFile(std::shared_ptr<OpenFileDescriptor> ptr);
+
+    // Removes OpenFileDescriptor in FileTable entries corresponding to given fd.
+    void removeOpenFile(__wasi_fd_t fd);
+
+    // Entry is used to override the subscript [] operator.
+    // This allows the user to get and set values in the FileTable entries vector.
+    // Syntax: std::shared_ptr<OpenFileDescriptor> openFile = currentHandle[fd];
+    struct Entry {
+      Handle& fileTableHandle;
+      __wasi_fd_t fd;
+      operator std::shared_ptr<OpenFileDescriptor> &() const {
+        assert(fd < fileTableHandle.size() && fd >= 0);
+
+        return fileTableHandle.fileTable.entries[fd];
+      }
```

Comment:
It would be good to have this method work for any value of `fd` since the other `Entry` methods work for any value of `fd` (or at least for positive values). I just realized that if we have this conversion return the `shared_ptr` by value rather than by reference, then we could return the `nullptr` value for invalid indices and wouldn't need `operator bool()` any more. Also, `operator->()` could also return the shared_ptr by value and could be implemented as `return std::shared_ptr<OpenFileDescriptor>(*this);` to avoid the same problem with not handling out-of-bounds fds.

Reply:
https://github.com/emscripten-core/emscripten/pull/15181#discussion_r718997062

> I wonder if this should return std::shared_ptr<OpenFile>& to avoid increasing the reference count unnecessarily. Same with the parameter type in add_open_file.

I think we had this originally returning a `nullptr` because it was changed because we didn't want the reference count to increase. So does this make sense to revert back to the original logic?